lib/builder.js

```diff
@@ -66,46 +67,75 @@
               }
             }
           } else {
-            for (key in obj) {
-              if (!hasProp.call(obj, key)) continue;
-              child = obj[key];
-              if (key === attrkey) {
-                if (typeof child === "object") {
-                  for (attr in child) {
-                    value = child[attr];
-                    element = element.att(attr, value);
-                  }
-                }
-              } else if (key === charkey) {
-                if (_this.options.cdata && requiresCDATA(child)) {
-                  element = element.raw(wrapCDATA(child));
-                } else {
-                  element = element.txt(child);
-                }
-              } else if (Array.isArray(child)) {
+            children_as_array = false;
+            if ((obj != null) && attrkey in obj && typeof obj[attrkey] === "object") {
+              child = obj[attrkey];
+              for (attr in child) {
+                value = child[attr];
+                element = element.att(attr, value);
+              }
+            }
+            if ((obj != null) && charkey in obj) {
+              child = obj[charkey];
+              if (_this.options.cdata && requiresCDATA(child)) {
+                element = element.raw(wrapCDATA(child));
+              } else {
+                element = element.txt(child);
+              }
+            }
+            if ((obj != null) && childkey in obj) {
+              child = obj[childkey];
+              if (Array.isArray(child)) {
+                children_as_array = true;
                 for (index in child) {
                   if (!hasProp.call(child, index)) continue;
                   entry = child[index];
-                  if (typeof entry === 'string') {
-                    if (_this.options.cdata && requiresCDATA(entry)) {
-                      element = element.ele(key).raw(wrapCDATA(entry)).up();
+                  if (typeof entry === "object") {
+                    if (Object.keys(entry).length === 1) {
+                      name = Object.keys(entry)[0];
+                      element = render(element.ele(name), entry[name]).up();
```

Comment:
This throws an error for self-closing elements without attributes because `entry` for `<foo />` looks like:
```json
{
  "#name": "foo"
}
```

I think `if ('#name' in entry)` should be checked first.

Can you add round-trip test cases for self-closing elements, both with and without attributes?